index.test.js

```diff
@@ -31,10 +49,356 @@ test('hoists at-root', () => {
   run('a { & {} @at-root { b {} } }', 'a {} b {}')
 })
 
+test('hoists at-root 2', () => {
+  run('a { @at-root { b {} } }', 'b {}')
+})
+
 test('at-root short hand', () => {
   run('a { & {} @at-root b { } }', 'a {} b {}')
 })
 
+test('hoists multiple at-roots', () => {
+  run(
+    `a {
+      b {
+        & {}
+        @at-root {
+          c1 {}
+          c2 {}
+        }
+        @at-root {
+          d {}
+        }
+      }
+    }`,
+    `a b {}
+    c1 {}
+    c2 {}
+    d {}`
+  )
+})
+
+test('hoists at-root and media siblings', () => {
+  run(
+    `a {
+      x: x;
+      a2 {}
+      @at-root {
+        b {}
+      }
+      @media x {
+        c {}
+      }
+      /* asdadf */
+    }`,
+    `a {
+      x: x;
+      /* asdadf */
+    }
+    a a2 {}
+    b {}
+    @media x {
+      a c {}
+    }`
+  )
+})
+
+test('at-root stops at media', () => {
+  run('@media x { a { & {} @at-root { b { } } } }', '@media x { a {} b {} }')
+})
+
+test('at-root unwraps nested media', () => {
+  run('a { & {} @media x { @at-root { b { } } } }', 'a {} @media x { b {} }')
+})
+
+test('nested at-root with nested media', () => {
+  run(
+    `a {
+      & {}
+      @at-root {
+      b {
+        @at-root {
+          c {
+            & {}
+          }
+          @media y {
+            d {}
+          }
+        }
+      }}
+    }`,
+    `a {}
+    c {}
+    @media y {
+      d {}
+    }`
+  )
+})
+
+test('at-root supports (without: all)', () => {
+  run(
+    `@media x {
+      @supports (z:y) {
+        a {
+          & {}
+          @at-root (without: all) {
+            b {}
+            @media y {
+              c {}
+            }
+          }
+          b {}
+        }
+      }
+    }`,
+    /*
+      NOTE: This is different from how SCSS does this.
+      SCSS moves the @at-root breakout block after/below the rest of
+      the subsequent sibling rules, which can sometimes cause quite
+      surprising results.
+
+      Since postcss-nested already does not perfectly emulate SCSS
+      (no pointless @media block splitting and query merging) it seems
+      justifiable to be "better than SCSS" in this case.
+    */
```

Comment:
I'm tempted to leave this comment in place. It serves as a justification for why/how this plugin behaves differently from SCSS.

We could later add a configuration option `rootRuleBehavior: 'logical' | 'scss',` to allow users to opt into closer compatibility with SCSS. This, however, feels kind of pointless as the `@media` block hanling is already different.

Reply:
I don’t think we need compatibility with SCSS. Anyway, there are too many differences.

Reply:
Agreed. I removed these comments.